File: src/common/py/core/messaging/composers/message_composer.py
import abc
import typing

from .. import (
    Message,
    MessageType,
    MessageBusProtocol,
    Channel,
    Payload,
    PayloadData,
)
from ..meta import MessageMetaInformation
from ....utils import UnitID

BeforeDispatchCallback = typing.Callable[[Message, MessageMetaInformation], None]


class MessageComposer(abc.ABC):
    """
    A class to collect all information necessary to create and emit a message.
    """

    def __init__(
        self,
        origin_id: UnitID,
        message_bus: MessageBusProtocol,
        msg_type: type[MessageType],
        chain: Message | None = None,
        **kwargs
    ):
        """
        Args:
            origin_id: The component identifier of the origin of newly created messages.
            message_bus: The global message bus to use.
            msg_type: The message type.
            chain: A message that acts as the *predecessor* of the new message. Used to keep the same trace for multiple messages.
            **kwargs: Additional message parameters.
        """
        if chain is not None:
            kwargs["trace"] = chain.trace

        self._origin_id = origin_id
        self._message_bus = message_bus

        self._msg_type = msg_type
        self._params = kwargs
        self._payload: Payload = {}

        self._before_callbacks: typing.List[BeforeDispatchCallback] = []

    def add_payload(self, key: str, data: PayloadData) -> None:
        """
        Adds a payload item to the message.

        Args:
            key: The item key.
            data: The item data.
        """
        self._payload[key] = data

    def before(self, callback: BeforeDispatchCallback) -> typing.Self:
        """
        Adds a callback that will be called immediately before the message is dispatched.

        Args:
            callback: The callback to add.

        Returns:
            This composer instance to allow call chaining.
        """
        self._before_callbacks.append(callback)
        return self

    def emit(self, target: Channel, *, suppress_logging: bool = False) -> None:
        """
        Sends the built message through the message bus.

        Args:
            target: The target of the message.
            suppress_logging: Suppresses debug logging.
        """
        self._verify()
        self._compose()

        msg_meta = self._create_meta_information(suppress_logging)
        msg = self._create_message(target)

        for callback in self._before_callbacks:
            callback(msg, msg_meta)

        self._message_bus.dispatch(msg, msg_meta)

    def _verify(self) -> None:
        pass

    def _compose(self) -> None:
        pass

    @abc.abstractmethod
    def _create_meta_information(
        self, suppress_logging: bool
    ) -> MessageMetaInformation: ...

    def _create_message(self, target: Channel) -> MessageType:
        msg = typing.cast(
            Message,
            self._msg_type(
                origin=self._origin_id,
                sender=self._origin_id,
                target=target,
                hops=[self._origin_id],
                **self._params,
            ),
        )
        msg.payload.decode(self._payload)
        return typing.cast(MessageType, msg)

File: src/common/web/core/messaging/composers/MessageComposer.ts
import { UnitID } from "../../../utils/UnitID";
import { Channel } from "../Channel";
import { type ConstructableMessage, Message } from "../Message";
import { type MessageBusProtocol } from "../MessageBusProtocol";
import { type Payload, PayloadData } from "../MessagePayload";
import { MessageMetaInformation } from "../meta/MessageMetaInformation";

export type BeforeDispatchCallback = (msg: Message, msgMeta: MessageMetaInformation) => void;

/**
 * A class to collect all information necessary to create and emit a message.
 */
export abstract class MessageComposer<MsgType extends Message> {
    protected _originID: UnitID;
    protected _messageBus: MessageBusProtocol;

    protected _msgType: ConstructableMessage<MsgType>;
    protected _params: Record<string, any>;
    protected _payload: Payload = {};
    protected _chain: Message | null;

    private _beforeCallbacks: BeforeDispatchCallback[] = [];

    /**
     * @param originID - The component identifier of the origin of newly created messages.
     * @param messageBus - The global message bus to use.
     * @param msgType - The message type.
     * @param params - Additional message parameters.
     * @param chain - A message that acts as the *predecessor* of the new message. Used to keep the same trace for multiple messages.
     */
    public constructor(
        originID: UnitID,
        messageBus: MessageBusProtocol,
        msgType: ConstructableMessage<MsgType>,
        params: Record<string, any> = {},
        chain: Message | null = null
    ) {
        this._originID = originID;
        this._messageBus = messageBus;

        this._msgType = msgType;
        this._params = params;
        this._chain = chain;
    }

    /**
     * Adds a payload item to the message.
     *
     * @param key - The item key.
     * @param data - The item data.
     */
    public addPayload(key: string, data: PayloadData): void {
        this._payload[key] = data;
    }

    /**
     * Adds a callback that will be called immediately before the message is dispatched.
     *
     * @param cb - The callback to add.
     *
     * @returns - This composer instance to allow call chaining.
     */
    public before(cb: BeforeDispatchCallback): this {
        this._beforeCallbacks.push(cb);
        return this;
    }

    /**
     * Sends the built message through the message bus.
     *
     * @param target - The target of the message.
     */
    public emit(target: Channel): void {
        this.verify();
        this.compose();

        let msgMeta = this.createMetaInformation();
        let msg = this.createMessage(target);

        for (const cb of this._beforeCallbacks) {
            cb(msg, msgMeta);
        }

        this._messageBus.dispatch(msg, msgMeta);
    }

    protected verify(): void {}

    protected compose(): void {}

    protected abstract createMetaInformation(): MessageMetaInformation;

    private createMessage(target: Channel): Message {
        let msg = this._chain
            ? new this._msgType(this._originID, this._originID, target, [this._originID], this._chain.trace)
            : new this._msgType(this._originID, this._originID, target, [this._originID]);

        for (const [key, value] of Object.entries(this._params)) {
            if (key in msg) {
                msg[key as keyof MsgType] = value;
            } else {
                throw new Error(`Provided unknown value '${key}' for message type ${msg.name}`);
            }
        }
        msg.payload.decode(this._payload);
        return msg;
    }
}
